### DISCUSSION TITLE
how to make a route+handler that accepts an optional value parameter?

### DISCUSSION THREAD

**Discussion by dan-da** [tokio-rs/axum#2744] [Q&A] [answered]
So I have URLs that accept a selector field that may or may not accept another argument.  In this (contrived) example the second path element is the selector and the 3rd element (if present) is a value that only pertains to certain selectors:

```
        /person/ceo
        /person/cfo
        /person/id/45224
        /person/hash/fa234cad452cd45
```

And I have a `PersonSelector` struct that can accept any of these forms and convert it into a canonical `PersonIdentifier`.

note that there are basically two url forms:  `/person/selector` and `/person/selector/value`

I would like to define a single route for "/person" that can handle either variant, passing the selector and optional value into `PersonSelector`.   Alternatively, two routes that call a single handler fn.

I haven't figured out how to make that work.  So I resorted to defining two routes that call two separate handler fn, and then each handler calls a common worker fn.  This works perfectly, but seems rather inelegant, especially as the pattern is repeated a few times.

```
        .route("/person/:selector", get(person_page))
        .route("/person/:selector/:value", get(person_page_with_value))
```

So I guess my question boils down to:  is there a way to make the `person` route and `person_page` handler optionally accept a value?

**Comment by jplatte** [comment, marked as answer]
In general, I would recommend that you just have two handler function that delegate most of their work to the same private function that is never used as a handler directly. However, both of the things you are asking can be achieved to some extent:

- If you want to only call `.route` once to register a handler for _anything starting with `/person/`, you can do so with `.route("/person/*rest", get(handler))`; there is no syntax for limiting the number of trailing segments though
- If you want to extract either one or two path parameters with `axum::extract::Path<T>`, then `T` has to be a type that implements `Deserialize` such that both inputs work, for example you could use `Path<Vec<String>>` for an arbitrary number of parameters, or a custom type like this for the two routes shown in your example:
  ```rust
  #[derive(Deserialize)]
  struct PersonPathParams {
      selector: String,
      value: Option<String>,
  }
  ```

**Reply by dan-da** [reply]
that is very helpful, thank-you!    I haven't tried it yet, but it seems combining the two techniques should achieve the desired result.

**Comment by dan-da** [comment]
Just reporting back that I was able to combine these techniques and it is working perfectly with a single route and single handler for each page.   This keeps both the routes and handlers simpler for anyone reading the code.

thx again @jplatte for the suggestions.
